File: Perceptron.js
/*
    A rede neural implementada no código fornecido é um Perceptron de múltiplas camadas, 
    com uma única camada oculta. Especificamente, é um tipo de Perceptron conhecido como Perceptron de 
    Camada Única (Single-Layer Perceptron) com uma camada oculta adicional. Essa arquitetura é comumente 
    referida como uma rede neural feedforward de duas camadas.

    Aqui está uma breve descrição da arquitetura:

    Camada de Entrada: Esta é a camada que recebe os dados de entrada. Cada nó na camada de entrada 
    representa uma característica ou atributo dos dados de entrada.

    Camada Oculta: Esta camada é onde ocorre o processamento intermediário. Cada nó na camada oculta 
    recebe entradas da camada de entrada e calcula uma combinação linear das entradas ponderadas pelos pesos. 
    Em seguida, aplica uma função de ativação não linear para produzir as saídas da camada oculta.

    Camada de Saída: Esta camada produz as saídas finais da rede neural. Assim como na camada oculta, 
    cada nó na camada de saída calcula uma combinação linear das entradas ponderadas pelos pesos e aplica uma 
    função de ativação para gerar as saídas finais.

    Portanto, essa rede neural em particular é um Perceptron de duas camadas, onde a camada oculta 
    introduz uma etapa intermediária de processamento que pode capturar relações não lineares nos dados, 
    tornando-a mais flexível do que um Perceptron de camada única.
*/
class Perceptron {
    #propability = 0.9;
 
     constructor(inputSize, bias = 0.1) {
         this.weights = new Array(inputSize);
         this.bias = bias; //Bias é a taxa de aprendizado -> Por padrão é 0.1
 
         // Inicializando os pesos aleatoriamente entre -1 e 1, incluindo o bias
         for (let i = 0; i < inputSize + 1; i++) {
             this.weights[i] = Math.random() * 2 - 1;
         }
     }
 
     // Função de ativação (degrau)
     activation(input) {
         return input >= 0 ? 1 : 0;
     }
 
     // Função para fazer previsões
     predict(inputs) {
         this.#validateInputs(inputs);
 
         let sum = this.weights[0]; // Incluindo o bias
         for (let i = 0; i < inputs.length; i++) {
             sum += inputs[i] * this.weights[i + 1]; // Os pesos começam do índice 1
         }
 
         const output = this.activation(sum);
         return output;
     }
 
     calculateWeightDeltas(error, output) {
         return error * output * (1 - output); // Gradiente da função de ativação (derivada da função de ativação sigmoid)
     }
 
     // Calcular o gradiente para a camada oculta
     calculateHiddenErrors(inputs, delta) {
         const hiddenLayerGradient = [];
         for (let i = 0; i < inputs.length; i++) {
             hiddenLayerGradient[i] = delta * this.weights[i + 1];
         }
 
         return hiddenLayerGradient;
     }
 
     backpropagation(inputs, target) {
         const output = this.predict(inputs);
         const error = target - output;
         const delta = this.calculateWeightDeltas(error, output);
         const learningRate = this.#propability;
     
         // Atualizar pesos da camada de saída
         this.weights[0] += delta * this.bias * learningRate; // Ajuste para o bias
         for (let i = 0; i < inputs.length; i++) {
             this.weights[i + 1] += delta * inputs[i] * this.bias * learningRate; // Ajustes para os pesos
         }
     
         const hiddenLayerGradient = this.calculateHiddenErrors(inputs, delta);
     
         // Atualizar pesos da camada oculta
         for (let i = 0; i < inputs.length; i++) {
             this.weights[i + 1] += hiddenLayerGradient[i] * inputs[i] * (1 - inputs[i]) * this.bias;
         }
     }
     
 
     #validateInputs(inputs) {
         if (inputs.length !== this.weights.length - 1) {
             throw new Error('O número de inputs não corresponde ao número de pesos.');
         }
     }
 
     //Função de treinamento
     train(inputs, target) {
         const guess = this.predict(inputs);
         const error = target - guess;
 
         // Ajustando os pesos com base no erro, incluindo o bias
         this.weights[0] += error * this.bias; // Ajuste para o bias
         for (let i = 0; i < inputs.length; i++) {
             this.weights[i + 1] += error * inputs[i] * this.bias; // Ajustes para os pesos
         }
 
         this.backpropagation(inputs, target);
     }
 }
 
 // Exemplo de uso
 // const perceptron = new Perceptron(2); // Criando um perceptron com 2 inputs
 // const trainingData = [
 //     { inputs: [0, 0], target: 0 },
 //     { inputs: [1, 0], target: 1 },
 //     { inputs: [0, 1], target: 1 },
 //     { inputs: [1, 1], target: 1 }
 // ];
 
 // // Treinando o perceptron
 // for (let i = 0; i < 10000; i++) {
 //     const data = trainingData[Math.floor(Math.random() * trainingData.length)];
 //     perceptron.train(data.inputs, data.target);
 // }
 
 // // Fazendo previsões
 // console.log(perceptron.predict([0, 0])); // Saída esperada: 0
 // console.log(perceptron.predict([1, 0])); // Saída esperada: 1
 // console.log(perceptron.predict([0, 1])); // Saída esperada: 1
 // console.log(perceptron.predict([1, 1])); // Saída esperada: 1
 
 
 //Target 0 - Cão
 //Target 1 - Gato
 
 //Input 0 - Peso
 //Input 1 - Altura
 //Input 2 - Orelha pontuda
 //input 3 - Bigode
 //Input 4 - Pupila vertical
 
 const perceptron = new Perceptron(5);
 
 const trainingData = [
     { inputs: [1, 1, 1, 0, 0], target: 0 },   // Cão
     { inputs: [1, 0, 0, 0, 0], target: 0 },   // Cão
     { inputs: [0, 0, 0, 0, 0], target: 0 },   // Cão
     { inputs: [1, 0, 0, 1, 1], target: 1 },   // Gato
     { inputs: [1, 1, 1, 1, 1], target: 1 },   // Gato
     { inputs: [0, 0, 0, 1, 1], target: 1 },   // Gato
     { inputs: [0, 0, 0, 0, 1], target: 1 },   // Gato
 ];
 
 // Treinando o perceptron
 for (let i = 0; i < 10000; i++) {
     const data = trainingData[Math.floor(Math.random() * trainingData.length)];
     perceptron.train(data.inputs, data.target);
 }
 
 const translate = {
     0: "Cão",
     1: "Gato"
 };
 
 // Fazendo previsões
 // console.log(translate[perceptron.predict([0, 0, 0, 0, 0])]); //Cão
 // console.log(translate[perceptron.predict([1, 0, 0, 1, 1])]); //Gato
 // console.log(translate[perceptron.predict([1, 1, 1, 0, 0])]); //Cão
 // console.log(translate[perceptron.predict([0, 0, 0, 1, 1])]); //Gato
 
 console.log(translate[perceptron.predict([0, 0, 1, 0, 0])]); //Cão
 

File: Perceptron.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#include <time.h>

#define PROPABILITY 0.9
#define BIAS 0.1

struct Neuron {
    double *weights;
    double bias;
};

void constructNeuron(struct Neuron *neuron, int inputSinapse) {
    neuron->weights = (double *)malloc(inputSinapse * sizeof(double));
    neuron->bias = BIAS;
}

void initializeWeights(struct Neuron *neuron, int inputSinapse) {
    for (int i = 0; i < inputSinapse + 1; i++) {
        neuron->weights[i] = ((double)rand() / RAND_MAX) * 2 - 1;
    }
}

int activation(double input) {
    return input >= 0 ? 1 : 0;
}

int sigmoid(double input) {
    return 1 / (1 + exp(-input));
}

int output(struct Neuron *neuron, const double *inputs, bool isTraining) {
    double sum = neuron->weights[0];

    for (int i = 0; i < 5; i++) {
        sum += inputs[i] * neuron->weights[i + 1];
    }

    if (isTraining) {
        return sigmoid(sum);
    }

    return activation(sum);
}

double calculateWeightDeltas(double error, double output) {
    return error * output * (1 - output);
}

double *calculateHiddenErrors(struct Neuron *neuron, const double *inputs, double delta) {
    double *hiddenLayerGradient = (double *)malloc(5 * sizeof(double));

    for (int i = 0; i < 5; i++) {
        hiddenLayerGradient[i] = delta * neuron->weights[i + 1];
    }

    return hiddenLayerGradient;
}

void updateWeights(struct Neuron *neuron, double delta, const double *inputs) {
    double learningRate = PROPABILITY;

    neuron->weights[0] += delta * neuron->bias * learningRate;
    for (int i = 0; i < 5; i++) {
        neuron->weights[i + 1] += delta * inputs[i] * neuron->bias * learningRate;
    }
}

void updateWeightsHiddenLayer(struct Neuron *neuron, const double *inputs, double *hiddenLayerGradient) {
    for (int i = 0; i < 5; i++) {
        neuron->weights[i + 1] += inputs[i] * (1 - inputs[i]) * neuron->bias;
    }
}

void backPropagation(struct Neuron *neuron, const double *inputs, int target) {
    double outputValue = output(neuron, inputs, true);
    double error = target - outputValue;
    double delta = calculateWeightDeltas(error, outputValue);

    updateWeights(neuron, delta, inputs);

    double *hiddenLayerGradient = calculateHiddenErrors(neuron, inputs, delta);

    updateWeightsHiddenLayer(neuron, inputs, hiddenLayerGradient);
    free(hiddenLayerGradient);
}

int main() {
    srand(time(NULL)); // Seed de inicialização para valores aleatórios

    struct Neuron perceptron;
    constructNeuron(&perceptron, 5);

    const int targetValues[] = {0, 0, 0, 0, 1, 1, 1, 1};
    const double trainingData[][5] = {
        {1, 1, 1, 0, 0},
        {1, 0, 0, 0, 0},
        {0, 0, 0, 0, 0},
        {0, 0, 1, 0, 0},
        {1, 0, 0, 1, 1},
        {1, 1, 1, 1, 1},
        {0, 0, 0, 1, 1},
        {0, 0, 0, 0, 1}
    };

    for (int i = 0; i < 10000; i++) {
        int randomIndex = rand() % 7;
        backPropagation(&perceptron, trainingData[randomIndex], targetValues[randomIndex]);
    }

    const char *translate[] = {"Cão", "Gato"};

    double pre = output(&perceptron, (double[]){1, 1, 0, 0, 0}, false);
    double pre2 = output(&perceptron, (double[]){1, 1, 1, 1, 1}, false);

    printf("Previsão: %s\n", translate[(int)pre]);
    printf("Previsão: %s\n", translate[(int)pre2]);

    free(perceptron.weights);
    return 0;
}